### ISSUE TITLE
[BUG] Changes to handlers using useState are ignored

### ISSUE THREAD

**Issue by joepio** [greena13/react-hotkeys#182]
**Describe the bug**
Changes in my functional component state (using React Hooks) are not properly reflected by the handlers, even when I've set `allowChanges` to true.

In practice, this means that my `handleNextPage` action uses an old version of the `currentPage` prop, which is set using `React.useState`. So the first time I want to go to the next page (e.g. page 2), it works fine, but the next time, it uses the old state (goes to page 2 again).

Whenever I re-focus on the HotKeys component (ref), the props are adjusted correctly and everything works as it should.

The app can be viewed [here](http://ori.argu.co/?searchbox=%22omgevingswet%22&showResource=https%253A%252F%252Fid.openraadsinformatie.nl%252F66673) and the relevant source code is [here](https://github.com/ontola/ori-search/blob/master/front/src/Components/PDFViewer.tsx#L171).

**How are you using react hotkeys components? (HotKeys, GlobalHotKeys, IgnoreKeys etc)**
I'm using only the HotKeys component, none of the others.

**Expected behavior**
I expect the HotKeys component to update when the `handler` functions change.

**Platform (please complete the following information):**
 - Version of react-hotkeys: ^2.0.0-pre5
 - Browser Chrome, latest
 - OS: Ubuntu 18.04

**Are you willing and able to create a PR request to fix this issue?**
If I can solve it, I definitely will.

**APPLICABLE TO v2.0.0-pre1 AND ABOVE: ======================**

**Include the smallest log that includes your issue:**

```
HotKeys (F0📕-E0❤️-C0🔺) Registered keyMap:
 {
    "PREVIOUS": "left",
    "NEXT": "right",
    "SEARCH": "/"
}
AbstractKeyEventStrategy.js:391 HotKeys (F0📕-E0❤️-C0🔺) Registered component:
 {
    "childIds": [],
    "parentId": null
}
AbstractKeyEventStrategy.js:422 HotKeys (F0📕-E0❤️-C0🔺) Registered component mount:
 {
    "childIds": [],
    "parentId": null
}
AbstractKeyEventStrategy.js:389 HotKeys (F0📕-E0❤️-C1⭐️) Registered keyMap:
 {}
AbstractKeyEventStrategy.js:391 HotKeys (F0📕-E0❤️-C1⭐️) Registered component:
 {
    "childIds": [],
    "parentId": null
}
AbstractKeyEventStrategy.js:422 HotKeys (F0📕-E0❤️-C1⭐️) Registered component mount:
 {
    "childIds": [],
    "parentId": 0
}
HotKeys (F0📕-C0🔺-P0🔺:) Focused. 

FocusOnlyKeyEventStrategy.js:308 HotKeys (F0📕-C0🔺-P0🔺:) Component options:
 {
    "actions": {
        "PREVIOUS": [
            {
                "prefix": "",
                "actionName": "PREVIOUS",
                "sequenceLength": 1,
                "id": "ArrowLeft",
                "keyDictionary": {
                    "ArrowLeft": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ],
        "NEXT": [
            {
                "prefix": "",
                "actionName": "NEXT",
                "sequenceLength": 1,
                "id": "ArrowRight",
                "keyDictionary": {
                    "ArrowRight": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ],
        "SEARCH": [
            {
                "prefix": "",
                "actionName": "SEARCH",
                "sequenceLength": 1,
                "id": "/",
                "keyDictionary": {
                    "/": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ]
    },
    "handlers": {
        "SEARCH": "function SEARCH() {\n    return console.log(\"search!\");\n  }"
    },
    "componentId": 0,
    "options": {
        "defaultKeyEvent": "keydown"
    }
}
FocusOnlyKeyEventStrategy.js:373 HotKeys (F0📕-C0🔺-P0🔺:) Lost focus.
FocusOnlyKeyEventStrategy.js:302 HotKeys (F1📗-C1⭐️-P0🔺:) Focused. 

FocusOnlyKeyEventStrategy.js:308 HotKeys (F1📗-C1⭐️-P0🔺:) Component options:
 {
    "actions": {},
    "handlers": {
        "PREVIOUS": "function handlePreviousPage() {\n    if (pageNumber === 1) {\n      return;\n    }\n\n    setPageNumber(pageNumber - 1);\n  }",
        "NEXT": "function handleNextPage() {\n    if (numPages === pageNumber) {\n      return;\n    }\n\n    setPageNumber(pageNumber + 1);\n  }"
    },
    "componentId": 1,
    "options": {
        "defaultKeyEvent": "keydown"
    }
}
FocusOnlyKeyEventStrategy.js:302 HotKeys (F1📗-C0🔺-P1⭐️:) Focused. 

FocusOnlyKeyEventStrategy.js:308 HotKeys (F1📗-C0🔺-P1⭐️:) Component options:
 {
    "actions": {
        "PREVIOUS": [
            {
                "prefix": "",
                "actionName": "PREVIOUS",
                "sequenceLength": 1,
                "id": "ArrowLeft",
                "keyDictionary": {
                    "ArrowLeft": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ],
        "NEXT": [
            {
                "prefix": "",
                "actionName": "NEXT",
                "sequenceLength": 1,
                "id": "ArrowRight",
                "keyDictionary": {
                    "ArrowRight": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ],
        "SEARCH": [
            {
                "prefix": "",
                "actionName": "SEARCH",
                "sequenceLength": 1,
                "id": "/",
                "keyDictionary": {
                    "/": true
                },
                "eventBitmapIndex": 0,
                "size": 1
            }
        ]
    },
    "handlers": {
        "SEARCH": "function SEARCH() {\n    return console.log(\"search!\");\n  }"
    },
    "componentId": 0,
    "options": {
        "defaultKeyEvent": "keydown"
    }
}
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) New 'ArrowRight' keydown event.
FocusOnlyKeyEventStrategy.js:744 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Added 'ArrowRight' to current combination: 'ArrowRight'.
FocusOnlyKeyEventStrategy.js:749 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    false,
                    false,
                    false
                ],
                [
                    true,
                    false,
                    false
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:874 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Attempting to find action matching 'ArrowRight' keydown . . .
AbstractKeyEventStrategy.js:975 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Internal key mapping:
 {
    "sequences": {
        "": {
            "combinations": {
                "ArrowLeft": {
                    "prefix": "",
                    "sequenceLength": 1,
                    "id": "ArrowLeft",
                    "keyDictionary": {
                        "ArrowLeft": true
                    },
                    "size": 1,
                    "events": {
                        "0": {
                            "actionName": "PREVIOUS",
                            "eventBitmapIndex": 0,
                            "handler": "function handlePreviousPage() {\n    if (pageNumber === 1) {\n      return;\n    }\n\n    setPageNumber(pageNumber - 1);\n  }"
                        }
                    }
                },
                "ArrowRight": {
                    "prefix": "",
                    "sequenceLength": 1,
                    "id": "ArrowRight",
                    "keyDictionary": {
                        "ArrowRight": true
                    },
                    "size": 1,
                    "events": {
                        "0": {
                            "actionName": "NEXT",
                            "eventBitmapIndex": 0,
                            "handler": "function handleNextPage() {\n    if (numPages === pageNumber) {\n      return;\n    }\n\n    setPageNumber(pageNumber + 1);\n  }"
                        }
                    }
                }
            }
        }
    },
    "eventBitmap": [
        true,
        false,
        false
    ],
    "longestSequence": 1
}
AbstractKeyEventStrategy.js:1036 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Found action that matches 'ArrowRight' (sub-match: 'ArrowRight'): NEXT. Calling handler . . .
FocusOnlyKeyEventStrategy.js:762 HotKeys (F1📗-E1💚-C0🔺-P1⭐️:) Stopping further event propagation.
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E3💛-C0🔺-P1⭐️:) New (simulated) 'ArrowRight' keypress event.
FocusOnlyKeyEventStrategy.js:749 HotKeys (F1📗-E3💛-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    false,
                    false
                ],
                [
                    true,
                    true,
                    false
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:889 HotKeys (F1📗-E3💛-C0🔺-P1⭐️:) Ignored 'ArrowRight' keypress because it doesn't have any keypress handlers.
FocusOnlyKeyEventStrategy.js:341 HotKeys (F1📗-C1⭐️-P0🔺:) Received new props.
FocusOnlyKeyEventStrategy.js:348 HotKeys (F1📗-C1⭐️-P0🔺:) Component options:
 {
    "actions": {},
    "handlers": {
        "PREVIOUS": "function handlePreviousPage() {\n    if (pageNumber === 1) {\n      return;\n    }\n\n    setPageNumber(pageNumber - 1);\n  }",
        "NEXT": "function handleNextPage() {\n    if (numPages === pageNumber) {\n      return;\n    }\n\n    setPageNumber(pageNumber + 1);\n  }"
    },
    "componentId": 1,
    "options": {
        "defaultKeyEvent": "keydown"
    }
}
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E4💜-C0🔺-P1⭐️:) New 'ArrowRight' keyup event.
FocusOnlyKeyEventStrategy.js:749 HotKeys (F1📗-E4💜-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    true,
                    false
                ],
                [
                    true,
                    true,
                    true
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:889 HotKeys (F1📗-E4💜-C0🔺-P1⭐️:) Ignored 'ArrowRight' keyup because it doesn't have any keyup handlers.
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) New 'ArrowRight' keydown event.
FocusOnlyKeyEventStrategy.js:731 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Started a new combination with 'ArrowRight'.
FocusOnlyKeyEventStrategy.js:734 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    true,
                    false
                ],
                [
                    true,
                    true,
                    true
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    },
    {
        "keys": {
            "ArrowRight": [
                [
                    false,
                    false,
                    false
                ],
                [
                    true,
                    false,
                    false
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:874 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Attempting to find action matching 'ArrowRight' keydown . . .
AbstractKeyEventStrategy.js:975 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Internal key mapping:
 {
    "sequences": {
        "": {
            "combinations": {
                "ArrowLeft": {
                    "prefix": "",
                    "sequenceLength": 1,
                    "id": "ArrowLeft",
                    "keyDictionary": {
                        "ArrowLeft": true
                    },
                    "size": 1,
                    "events": {
                        "0": {
                            "actionName": "PREVIOUS",
                            "eventBitmapIndex": 0,
                            "handler": "function handlePreviousPage() {\n    if (pageNumber === 1) {\n      return;\n    }\n\n    setPageNumber(pageNumber - 1);\n  }"
                        }
                    }
                },
                "ArrowRight": {
                    "prefix": "",
                    "sequenceLength": 1,
                    "id": "ArrowRight",
                    "keyDictionary": {
                        "ArrowRight": true
                    },
                    "size": 1,
                    "events": {
                        "0": {
                            "actionName": "NEXT",
                            "eventBitmapIndex": 0,
                            "handler": "function handleNextPage() {\n    if (numPages === pageNumber) {\n      return;\n    }\n\n    setPageNumber(pageNumber + 1);\n  }"
                        }
                    }
                }
            },
            "order": [
                "ArrowLeft",
                "ArrowRight"
            ]
        }
    },
    "eventBitmap": [
        true,
        false,
        false
    ],
    "longestSequence": 1
}
AbstractKeyEventStrategy.js:1036 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Found action that matches 'ArrowRight' (sub-match: 'ArrowRight'): NEXT. Calling handler . . .
FocusOnlyKeyEventStrategy.js:762 HotKeys (F1📗-E5🧡-C0🔺-P1⭐️:) Stopping further event propagation.
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E7💚-C0🔺-P1⭐️:) New (simulated) 'ArrowRight' keypress event.
FocusOnlyKeyEventStrategy.js:749 HotKeys (F1📗-E7💚-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    true,
                    false
                ],
                [
                    true,
                    true,
                    true
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    },
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    false,
                    false
                ],
                [
                    true,
                    true,
                    false
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:889 HotKeys (F1📗-E7💚-C0🔺-P1⭐️:) Ignored 'ArrowRight' keypress because it doesn't have any keypress handlers.
FocusOnlyKeyEventStrategy.js:470 HotKeys (F1📗-E8💙-C0🔺-P1⭐️:) New 'ArrowRight' keyup event.
FocusOnlyKeyEventStrategy.js:749 HotKeys (F1📗-E8💙-C0🔺-P1⭐️:) Key history: [
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    true,
                    false
                ],
                [
                    true,
                    true,
                    true
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    },
    {
        "keys": {
            "ArrowRight": [
                [
                    true,
                    true,
                    false
                ],
                [
                    true,
                    true,
                    true
                ]
            ]
        },
        "ids": [
            "ArrowRight"
        ],
        "keyAliases": {}
    }
].
FocusOnlyKeyEventStrategy.js:889 HotKeys (F1📗-E8💙-C0🔺-P1⭐️:) Ignored 'ArrowRight' keyup because it doesn't have any keyup handlers.
FocusOnlyKeyEventStrategy.js:373 HotKeys (F1📗-C1⭐️-P0🔺:) Lost focus.
FocusOnlyKeyEventStrategy.js:373 HotKeys (F1📗-C0🔺-P1⭐️:) Lost focus.
KeyEventManager.js:183 HotKeys: Window focused - clearing key history

```

**What Configuration options are you using?**

```
configure({
  logLevel: "verbose",
})
```

**Comment by janKir** [comment]
I face the same issue, when trying to increment/decrement an index variable. It seems like changes to the index are not detected in the hotkey handler's closure so it always uses the initial value. However, the same handlers work e.g. with buttons. I use the `useState` hook if that's relevant here.

**Comment by greena13** [comment]
Hi @joepio,

Thanks for the great write-up of your issue and including a full log. This has been very helpful in understanding what's going on.

I don't have any experience with the [`useState()` function](https://reactjs.org/docs/hooks-state.html) introduced in React 16.8, so this may be new territory for `react-hotkeys`.

I can see from your log that `react-hotkeys` appears to be performing correctly, registering the new prop values and passing down the new handlers. 

A couple of things I'd like you to clarify:  the source code you linked to uses `GlobalHotKeys` (not `HotKeys`) and I can't see any reference to a `currentPage` prop. Did you make further changes or solve your problem since you posted this issue?

@janKir, thanks for letting me know multiple people are seeing this problem. The fact that you're also using `useState` has me rather curious.

**Comment by janKir** [comment]
I can confirm it's probably the `useState` hook causing problems. I converted my component into a stateful class component and now the HotKeys are working. But now other parts of my component don't work and need more effort this way, so this isn't a good workaround for me. Hope this can be fixed instead.

**Comment by joepio** [comment]
@greena13 

Thanks for the reply! I've tried both GlobalHotKeys and Hotkeys, the problem was the same. The name of the prop has been refactored from `curentPage` to `pageNumber`

**Comment by greena13** [comment]
Thanks for the clarification.

I've renamed the issue to match my current working theory and I'll try and dig into why it may be the case that `useState` may causing issues or not behaving in the manner that I would expect from a handler function.

Please let me know if you uncover anything that may help shed some light on the matter or guide my investigation.

**Comment by hiquest** [comment]
Not quite sure if need to put another "me too" comment here. But this is exactly what is happening to me when using useState and GlobalHotKeys. Hoping this will be fixed soon.

**Comment by greena13** [comment]
Hi @hiquest, "me too"s are always helpful to guage the prevelance and boundaries of the problem. Thanks for letting me know.

**Comment by janKir** [comment]
I found a workaround! I just realized that `this.setState()` can also be called with a function taking the old state as parameter. And [this can also be done with the `useState` hook](https://reactjs.org/docs/hooks-reference.html#functional-updates):

> If the new state is computed using the previous state, you can pass a function to setState. The function will receive the previous value, and return an updated value.

So, instead of referencing the state variable in the Hotkeys handler you can use the function parameter:

``` javascript
const [index, setIndex] = useState(0);

// in your handler:

// before - not working
setIndex(index + 1);

// after - working
setIndex(prevIndex => prevIndex + 1)
```

**Comment by greena13** [comment]
That's great news, @janKir. It should serve as short-term solution, while I have time to get to the bottom of this.

I have read up on [the useState() function](https://reactjs.org/docs/hooks-state.html) and the two [rules for when and how to use it](https://reactjs.org/docs/hooks-overview.html#rules-of-hooks):

> Only call Hooks at the top level. Don’t call Hooks inside loops, conditions, or nested functions.
>
> Only call Hooks from React function components. Don’t call Hooks from regular JavaScript functions.

At-a-glance, neither of these seem to violated by your application code.

I am still in the preliminary stages of investigating this, so I may very well have overlooked something. The next step is to upgrade the example app and have a play around with the new React Hooks API to see if I can reproduce.

**Comment by greena13** [comment]
I believe I've managed to track down the cause and craft a suitable fix (see linked to PR). It will go out with the next release.

**Comment by greena13** [comment]
Fix is now available in [v2.0.0-pre8](https://github.com/greena13/react-hotkeys/releases/tag/v2.0.0-pre8).
